test/pytest/test_beiboot.py

```diff
@@ -1,28 +1,459 @@
+import argparse
+import asyncio
+import json
+import os
+import subprocess
 import sys
+import typing
+from pathlib import Path
+from typing import Iterable
 
 import pytest
 
-from cockpit._vendor import ferny
-from cockpit._vendor.bei import bootloader
-from cockpit.beipack import BridgeBeibootHelper
-from cockpit.peer import Peer
-from cockpit.router import Router
+from cockpit.beiboot import Destination, ForwarderRouter, parse_destination
+from cockpit.jsonutil import JsonDocument, JsonObject, JsonValue, get_dict, get_str
 
+from .mocktransport import MockTransport
 
-class BeibootPeer(Peer):
-    async def do_connect_transport(self) -> None:
-        helper = BridgeBeibootHelper(self)
-        agent = ferny.InteractionAgent([helper])
-        transport = await self.spawn([sys.executable, '-iq'], env=[], stderr=agent)
-        transport.write(bootloader.make_bootloader(helper.steps, gadgets=ferny.BEIBOOT_GADGETS).encode())
-        await agent.communicate()
+if typing.TYPE_CHECKING:
+    import testvm
+else:
+    # Deeply weird: we want static checking, but dynamic loading
+    testvm = None
+
+
+LECTURE = '''
+We trust you have received the usual lecture from the local System
+Administrator. It usually boils down to these three things:
+
+    #1) Respect the privacy of others.
+    #2) Think before you type.
+    #3) With great power comes great responsibility.
+
+For security reasons, the password you type will not be visible.
+
+'''
+
+
+@pytest.fixture(scope='session', autouse=True)
+def bots() -> Path:
+    if not os.access('/dev/kvm', os.W_OK):
+        pytest.skip('kvm unavailable')
```

Comment:
ATM, pytest only runs on github workflows, so this test doesn't run anywhere.

So AFAICS we have two choices here:

 1. Treat this as unit test. Then this shouldn't involve VMs (or even containers) and network access. I get why we don't want to test against src/ssh/mock-sshd.c, as we are trying to get rid of that code and the libssh dependency. But ferny tests against [python-asyncssh](https://pypi.org/project/asyncssh/), and a [mock SSH server](https://github.com/allisonkarlitskaya/ferny/blob/main/test/test_basic.py#L109) looks quite reasonable?

2. Treat it as an integration test. Then this shouldn't reimplement all the bots/machine/hostkey/privkey handling here, but just use the bots API. You can [drive this from Python](https://github.com/cockpit-project/bots/blob/main/machine/machine_core/cli.py) in just a few lines of code if you insist on doing this with pytest fixtures (but honestly our standard `MachineCase` seems just fine for that, and it should be possible to wrap that into a fixture). The integration tests *already* assume that they run from the git checkout (e.g. as they need test/verify/files), so importing the beiboot module and running it on the host seems fine to me -- the integration tests don't *have to* involve a browser. E.g. most of test/verify/check-connection doesn't.

It makes sense to split the test. I.e. keep the "unit-y" bits here, like `test_parse_destination()` or `def test_conds()` (with some adjustments), and move the integration-y ones to test/verify/check-client or a new test/verify/check-beiboot.

Reply:
This is sort of what I was hoping you'd look into...

This is definitely some kind of a "third type" of test.  fwiw, I've been running them locally and they're helping me tonnes (and they all currently pass).

Reply:
Summary from our discussion:
 - we keep the general shape, but move it into test/pytest-integration/
 - drop that bots fixture and just import testvm statically; integration tests just assume bots/
 - add a pytest -k mumble mumble select test/pytest-integration/* to test/run, so that it runs on our CI independently of run-tests; it doesn't parallelize, but it's also fast enough (preferably the /networking scenario, it's the fastest)

But also: We don't want to/have to block this PR on that. We can live with not running these for a while. The more useful work would be to add a proper beiboot scenario on an unprepared image. The covered functionality is mostly just a faster/smaller version of what the integration tests already do.